File: src/main/java/com/sesame/worker/TextDocumentationWorkerImpl.java
package com.sesame.worker;

import com.sesame.core.worker.DocumentationWorker;
import com.sesame.domain.internal.DocMethod;
import com.sesame.domain.internal.ParameterType;
import com.sesame.util.TextUtils;
import com.sesame.worker.exception.DocumentInvalidFormatException;
import org.apache.commons.io.IOUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Optional;
import java.util.regex.PatternSyntaxException;

@Deprecated
public class TextDocumentationWorkerImpl implements DocumentationWorker {

    private final static Logger LOGGER = LoggerFactory.getLogger(TextDocumentationWorkerImpl.class);

    private static final String DELIMITATOR_COLON = ":";
    private static final String DELIMITATOR_DASH = "-";

    private static final String FIELD_BEAN_IDENTIFICATION = "beanIdentification";
    private static final String FIELD_PACKAGENAME = "packageName";
    private static final String FIELD_CLASSNAME = "className";
    private static final String FIELD_METHODNAME = "methodName";
    private static final String FIELD_METHODDESCRIPTION = "methodDescription";
    private static final String FIELD_RETURN = "return";
    private static final String FIELD_PARAM = "param";

    private enum Tag {
        BEAN_IDENTIFICATION_NAME, PACKAGE_NAME, CLASS_NAME, METHOD_NAME, METHOD_DESCRIPTION, RETURN, PARAM, OTHER
    }

    @Override
    public Optional<DocMethod> processInterfaceMethod(File file) {

        if (file == null) {
            return Optional.empty();
        }

        try {
            String configs = IOUtils.toString(Files.newInputStream(Paths.get(file.getAbsolutePath())), Charset.defaultCharset());


            Optional<String> filterDocumentation = TextUtils.extractDoclibTags(configs);

            if (!filterDocumentation.isPresent()) {
                return Optional.empty();
            }

            String[] docLines = filterDocumentation.get().split("\n");

            DocMethod.Builder docMethod = new DocMethod.Builder();

            for (String docLine : docLines) {

                String line = docLine.trim();

                try {
                    switch (getTag(line)) {
                        case BEAN_IDENTIFICATION_NAME:
                            String beanIdentification = line.split(DELIMITATOR_COLON)[1].trim();
                            docMethod.withBeanIdentification(beanIdentification);
                            break;
                        case PACKAGE_NAME:
                            String packageName = line.split(DELIMITATOR_COLON)[1].trim();
                            docMethod.withPackageName(packageName);
                            break;
                        case CLASS_NAME:
                            String className = line.split(DELIMITATOR_COLON)[1].trim();
                            docMethod.withClassName(className);
                            break;
                        case METHOD_NAME:
                            String methodName = line.split(DELIMITATOR_COLON)[1].trim();
                            docMethod.withMethodName(methodName);
                            break;
                        case METHOD_DESCRIPTION:
                            String methodDescrip = line.split(DELIMITATOR_COLON)[1].trim();
                            docMethod.withMethodDescription(methodDescrip);
                            break;
                        case RETURN:
                            String returnDocumentation = line.split(DELIMITATOR_COLON)[1].trim();

                            try {
                                String returnDescription = returnDocumentation.split(DELIMITATOR_DASH)[1].trim();
                                String returnType = returnDocumentation.split(DELIMITATOR_DASH)[0].trim();
                                docMethod.withReturnObject(returnType);
                                docMethod.withReturnObjectDescription(returnDescription);
                            } catch (PatternSyntaxException ex) {
                                LOGGER.warn("msg='Can not extract the value detail, check the format of documentation', return={}", returnDocumentation, ex);
                                docMethod.withReturnObject(returnDocumentation);
                            }

                            break;
                        case PARAM:
                            String paramDocumentation = line.split(DELIMITATOR_COLON)[1].trim();
                            int paramIndex = Integer.parseInt(line.split(DELIMITATOR_COLON)[0].split(DELIMITATOR_DASH)[1]);

                            try {
                                String paramDescription = paramDocumentation.split(DELIMITATOR_DASH)[1].trim();
                                String paramType = paramDocumentation.split(DELIMITATOR_DASH)[0].trim();
                                docMethod.addParamObjects(paramIndex, ParameterType.getInternalType(paramType));
                            } catch (PatternSyntaxException ex) {
                                LOGGER.warn("msg='Can not extract the param detail, check the format of documentation', param={}", paramDocumentation, ex);
                            }

                            break;
                        default:
                            break;
                    }
                } catch (Exception ex) {
                    LOGGER.error("msg='Can not extract the value, check the format of documentation', docLine={}", docLine, ex);
                    throw new DocumentInvalidFormatException("Error when try to extract the values form line: ", ex);
                }

            }

            return Optional.of(docMethod.build());

        } catch (IOException ex) {
            LOGGER.error("msg='Can not extract the value from config variable.'", ex);
            throw new DocumentInvalidFormatException("Error when try to extract the values form line: ", ex);
        }
    }

    private Tag getTag(String line) {

        if (line.startsWith(FIELD_METHODNAME)) {
            return Tag.METHOD_NAME;
        } else if (line.startsWith(FIELD_METHODDESCRIPTION)) {
            return Tag.METHOD_DESCRIPTION;
        } else if (line.startsWith(FIELD_RETURN)) {
            return Tag.RETURN;
        } else if (line.startsWith(FIELD_PARAM)) {
            return Tag.PARAM;
        } else if (line.startsWith(FIELD_CLASSNAME)) {
            return Tag.CLASS_NAME;
        } else if (line.startsWith(FIELD_PACKAGENAME)) {
            return Tag.PACKAGE_NAME;
        } else if (line.startsWith(FIELD_BEAN_IDENTIFICATION)) {
            return Tag.BEAN_IDENTIFICATION_NAME;
        }

        return Tag.OTHER;
    }


}

File: src/main/kotlin/com/sesame/core/worker/TextDocumentationWorkerImpl.kt
package com.sesame.core.worker

import com.sesame.core.domain.MethodInfo
import com.sesame.core.domain.ParameterType
import com.sesame.core.util.TextUtils
import org.apache.commons.io.IOUtils
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.io.File
import java.io.IOException
import java.nio.charset.Charset
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*
import java.util.regex.PatternSyntaxException

@Component
class TextDocumentationWorkerImpl : DocumentationWorker {

    companion object {
        private val LOGGER = LoggerFactory.getLogger(TextDocumentationWorkerImpl::class.java)

        private const val DELIMITATOR_COLON = ":"
        private const val DELIMITATOR_DASH = "-"

        private const val FIELD_BEAN_IDENTIFICATION = "beanIdentification"
        private const val FIELD_PACKAGE_NAME = "packageName"
        private const val FIELD_CLASSNAME = "className"
        private const val FIELD_METHOD_NAME = "methodName"
        private const val FIELD_METHOD_DESCRIPTION = "methodDescription"
        private const val FIELD_RETURN = "return"
        private const val FIELD_PARAM = "param"

    }

    private enum class Tag {
        BEAN_IDENTIFICATION_NAME,
        PACKAGE_NAME,
        CLASS_NAME,
        METHOD_NAME,
        METHOD_DESCRIPTION,
        RETURN,
        PARAM,
        OTHER
    }

    override fun processInterfaceMethod(file: File): Optional<MethodInfo> {

        try {

            val configs = IOUtils.toString(Files.newInputStream(Paths.get(file.absolutePath)), Charset.defaultCharset())

            val filterDocumentation = TextUtils.extractDocLibTags(configs)

            if (!filterDocumentation.isPresent) {
                return Optional.empty()
            }

            val docLines = filterDocumentation.get().split("\n".toRegex())

            val methodInfo = MethodInfo.Builder()

            for (docLine in docLines) {
                val line = docLine.trim()
                try {
                    when (getTag(line)) {
                        Tag.BEAN_IDENTIFICATION_NAME -> {
                            val beanIdentification = line.split(DELIMITATOR_COLON.toRegex())[1].trim()
                            methodInfo.withBeanIdentification(beanIdentification)
                        }

                        Tag.PACKAGE_NAME -> {
                            val packageName = line.split(DELIMITATOR_COLON.toRegex())[1].trim()
                            methodInfo.withPackageName(packageName)
                        }

                        Tag.CLASS_NAME -> {
                            val className = line.split(DELIMITATOR_COLON.toRegex())[1].trim()
                            methodInfo.withClassName(className)
                        }

                        Tag.METHOD_NAME -> {
                            val methodName = line.split(DELIMITATOR_COLON.toRegex())[1].trim()
                            methodInfo.withMethodName(methodName)
                        }

                        Tag.METHOD_DESCRIPTION -> {
                            val methodDescrip = line.split(DELIMITATOR_COLON.toRegex())[1].trim()
                            methodInfo.withMethodDescription(methodDescrip)
                        }

                        Tag.RETURN -> {
                            val returnDocumentation = line.split(DELIMITATOR_COLON.toRegex())[1].trim()
                            try {
                                val returnDescription = returnDocumentation.split(DELIMITATOR_DASH.toRegex())[1].trim()
                                val returnType = returnDocumentation.split(DELIMITATOR_DASH.toRegex())[0].trim()
                                methodInfo.withReturnObject(returnType)
                                methodInfo.withReturnObjectDescription(returnDescription)
                            } catch (ex: PatternSyntaxException) {
                                LOGGER.warn("message='Can not extract the value detail, check the format of documentation', return='{}'", returnDocumentation, ex)
                                methodInfo.withReturnObject(returnDocumentation)
                            }
                        }

                        Tag.PARAM -> {
                            val paramDocumentation = line.split(DELIMITATOR_COLON.toRegex())[1].trim()
                            val paramIndex = line.split(DELIMITATOR_COLON.toRegex())[0].split(DELIMITATOR_DASH.toRegex())[1].toInt()
                            try {
                                val paramDescription = paramDocumentation.split(DELIMITATOR_DASH.toRegex())[1].trim()
                                val paramType = paramDocumentation.split(DELIMITATOR_DASH.toRegex())[0].trim()
                                methodInfo.addParamObjects(paramIndex, ParameterType.getInternalType(paramType))
                            } catch (ex: PatternSyntaxException) {
                                LOGGER.warn("message='Can not extract the param detail, check the format of documentation', param='{}'", paramDocumentation, ex)
                            }
                        }
                        else -> {
                        }
                    }
                } catch (ex: Exception) {
                    LOGGER.error("message='Can not extract the value, check the format of documentation', docLine='{}'", docLine, ex)
                    throw RuntimeException("Error when try to extract the values from line")
                }
            }

            return Optional.of(methodInfo.build())

        } catch (ex: IOException) {
            LOGGER.error("message='Can not extract the value from config variable.'", ex)
            throw RuntimeException("Error when try to extract the values from line")
        }

    }

    private fun getTag(line: String): Tag {

        when {
            line.startsWith(FIELD_METHOD_NAME) -> {
                return Tag.METHOD_NAME
            }
            line.startsWith(FIELD_METHOD_DESCRIPTION) -> {
                return Tag.METHOD_DESCRIPTION
            }
            line.startsWith(FIELD_RETURN) -> {
                return Tag.RETURN
            }
            line.startsWith(FIELD_PARAM) -> {
                return Tag.PARAM
            }
            line.startsWith(FIELD_CLASSNAME) -> {
                return Tag.CLASS_NAME
            }
            line.startsWith(FIELD_PACKAGE_NAME) -> {
                return Tag.PACKAGE_NAME
            }
            line.startsWith(FIELD_BEAN_IDENTIFICATION) -> {
                return Tag.BEAN_IDENTIFICATION_NAME
            }
            else -> return Tag.OTHER
        }

    }

}